IdentityCore/src/oauth2/MSIDTokenResponseHandler.m

```diff
@@ -0,0 +1,172 @@
+                                     MSIDErrorInternal, @"Client info was not returned in the server response", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    // Checking for V1 response is done
+    if (![response isKindOfClass:MSIDAADV2TokenResponse.class] || ![parameters isKindOfClass:MSIDAADV2RequestParameters.class])
+    {
+        return YES;
+    }
+    
+    // The rest of the checking is only for V2
+    MSIDAADV2RequestParameters *aadV2Parameters = (MSIDAADV2RequestParameters *)parameters;
+    if ([NSString msidIsStringNilOrBlank:response.scope])
+    {
+        MSID_LOG_INFO(context, @"No scope in server response, using passed in scope instead.");
+        MSID_LOG_INFO_PII(context, @"No scope in server response, using passed in scope instead.");
+        [response setScope:aadV2Parameters.scopes.msidToString];
+    }
+    
+    // For silent flow, with grant type being MSID_OAUTH2_REFRESH_TOKEN, this value may be missing from the response.
+    // In this case, we simply return the refresh token in the request.
+    if (refreshToken)
+    {
+        if (!response.refreshToken)
+        {
+            response.refreshToken = refreshToken;
+            MSID_LOG_WARN(context, @"Refresh token was missing from the token refresh response, so the refresh token in the request is returned instead");
+            MSID_LOG_WARN_PII(context, @"Refresh token was missing from the token refresh response, so the refresh token in the request is returned instead");
+        }
+    }
+    
+    // TODO: ADAL and MSAL are checking if user matches in different places. Discuss if we should move that logic to this function
+    
+    return YES;
+}
+
++ (void)checkCorrelationId:(MSIDAADTokenResponse *)response
+      requestCorrelationId:(NSUUID *)requestCorrelationId
+{
+    MSID_LOG_VERBOSE_CORR(requestCorrelationId, @"Token extraction. Attempt to extract the data from the server response.");
+    
+    NSString *responseId = [response correlationId];
+    if (![NSString msidIsStringNilOrBlank:responseId])
+    {
+        NSUUID* responseUUID = [[NSUUID alloc] initWithUUIDString:responseId];
+        if (!responseUUID)
+        {
+            MSID_LOG_INFO_CORR(requestCorrelationId, @"Bad correlation id - The received correlation id is not a valid UUID. Sent: %@; Received: %@", requestCorrelationId, responseId);
+        }
+        else if (![requestCorrelationId isEqual:responseUUID])
+        {
+            MSID_LOG_INFO_CORR(requestCorrelationId, @"Correlation id mismatch - Mismatch between the sent correlation id and the received one. Sent: %@; Received: %@", requestCorrelationId, responseId);
+        }
+    }
+    else
+    {
+        MSID_LOG_INFO_CORR(requestCorrelationId, @"Missing correlation id - No correlation id received for request with correlation id: %@", [requestCorrelationId UUIDString]);
+    }
+}
+
++ (MSIDErrorCode)getErrorCodeForAADV2:(NSString *)oauthError
```

Comment:
as mentioned above this should be part of token response.

Reply:
done